packages/wrangler/e2e/pages-dev.test.ts

```diff
@@ -5,36 +5,87 @@ import getPort from "get-port";
 import shellac from "shellac";
 import { fetch } from "undici";
 import { beforeEach, describe, expect, it } from "vitest";
+import { normalizeOutput } from "./helpers/normalize";
 import { retry } from "./helpers/retry";
 import { dedent, makeRoot, seed } from "./helpers/setup";
 import { WRANGLER } from "./helpers/wrangler-command";
 
+type MaybePromise<T = void> = T | Promise<T>;
```

Comment:
technically, all of `MaybePromise`, `SessionData`, `waitUntilOutputContains` and `runDevSession` are duplicated from wrangler proper. The only diff is the `dev` command `runDevSession` runs under the hood (`wrangler dev` vs `wrangler pages dev`).

We could keep things as they are, in the idea that convergence will make one of these obsolete, or we could extract the common functionality in a helper to clean up the code duplication. No major blocker, just calling it out